File: server-rs/src/db/tasks.rs
use anyhow::{Context, Result};
use rusqlite::{OptionalExtension, params};
use serde_json::Value;
use uuid::Uuid;

use crate::models::{CreateTaskPayload, TaskWithPayload};

use super::{Db, UpsertTaskTransition, UpsertTasksResult, now_iso};

impl Db {
    pub fn upsert_tasks(
        &self,
        repo_id: &str,
        jira_account_id: &str,
        jira_board_id: &str,
        tasks: &[crate::provider::jira::JiraIssueForTask],
    ) -> Result<UpsertTasksResult> {
        if tasks.is_empty() {
            return Ok(UpsertTasksResult {
                synced: 0,
                transitions: Vec::new(),
            });
        }
        let mut conn = self.connect()?;
        let tx = conn.transaction()?;
        let ts = now_iso();
        let mut transitions = Vec::new();

        for task in tasks {
            let existing: Option<(String, String)> = tx
                .query_row(
                    "SELECT id, status FROM tasks WHERE jira_account_id = ?1 AND jira_issue_key = ?2",
                    params![jira_account_id, task.jira_issue_key],
                    |row| Ok((row.get(0)?, row.get(1)?)),
                )
                .optional()?;
            let task_id = existing
                .as_ref()
                .map(|(id, _)| id.clone())
                .unwrap_or_else(|| Uuid::new_v4().to_string());
            tx.execute(
                r#"INSERT INTO tasks (
                     id, repo_id, jira_account_id, jira_board_id, jira_issue_key, title,
                     description, assignee, status, priority, source, require_plan, auto_start,
                     auto_approve_plan, last_pipeline_error, last_pipeline_at, payload_json,
                     created_at, updated_at
                   ) VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, 'jira', 1, 0, 0, NULL, NULL, ?11, ?12, ?13)
                   ON CONFLICT(jira_account_id, jira_issue_key)
                   DO UPDATE SET
                     repo_id = excluded.repo_id,
                     jira_board_id = excluded.jira_board_id,
                     title = excluded.title,
                     description = excluded.description,
                     assignee = excluded.assignee,
                     priority = excluded.priority,
                     source = excluded.source,
                     payload_json = excluded.payload_json,
                     updated_at = excluded.updated_at"#,
                params![
                    task_id,
                    repo_id,
                    jira_account_id,
                    jira_board_id,
                    task.jira_issue_key,
                    task.title,
                    task.description,
                    task.assignee,
                    task.status,
                    task.priority,
                    task.payload.to_string(),
                    ts,
                    ts,
                ],
            )?;

            transitions.push(UpsertTaskTransition {
                task_id,
                is_new: existing.is_none(),
                previous_status: existing.as_ref().map(|(_, status)| status.clone()),
                current_status: existing
                    .as_ref()
                    .map(|(_, status)| status.clone())
                    .unwrap_or_else(|| task.status.clone()),
            });
        }
        tx.commit()?;
        Ok(UpsertTasksResult {
            synced: tasks.len(),
            transitions,
        })
    }

    pub fn list_tasks_by_repo(&self, repo_id: &str) -> Result<Vec<TaskWithPayload>> {
        let conn = self.connect()?;
        let mut stmt = conn.prepare(
            "SELECT id, repo_id, jira_account_id, jira_board_id, jira_issue_key, title, description, assignee, status, priority, source, require_plan, auto_start, auto_approve_plan, use_worktree, last_pipeline_error, last_pipeline_at, agent_profile_id, pr_url, pr_number, payload_json, created_at, updated_at FROM tasks WHERE repo_id = ?1 ORDER BY updated_at DESC",
        )?;
        let rows = stmt.query_map([repo_id], |row| {
            let payload_raw: String = row.get(20)?;
            Ok(TaskWithPayload {
                id: row.get(0)?,
                repo_id: row.get(1)?,
                jira_account_id: row.get(2)?,
                jira_board_id: row.get(3)?,
                jira_issue_key: row.get(4)?,
                title: row.get(5)?,
                description: row.get(6)?,
                assignee: row.get(7)?,
                status: row.get(8)?,
                priority: row.get(9)?,
                source: row.get(10)?,
                require_plan: row.get::<_, i64>(11)? != 0,
                auto_start: row.get::<_, i64>(12)? != 0,
                auto_approve_plan: row.get::<_, i64>(13)? != 0,
                use_worktree: row.get::<_, i64>(14)? != 0,
                last_pipeline_error: row.get(15)?,
                last_pipeline_at: row.get(16)?,
                agent_profile_id: row.get(17)?,
                pr_url: row.get(18)?,
                pr_number: row.get(19)?,
                payload: serde_json::from_str(&payload_raw).unwrap_or(Value::Null),
                created_at: row.get(21)?,
                updated_at: row.get(22)?,
            })
        })?;
        rows.collect::<std::result::Result<Vec<_>, _>>()
            .map_err(anyhow::Error::from)
    }

    pub fn get_task_by_id(&self, task_id: &str) -> Result<Option<TaskWithPayload>> {
        let conn = self.connect()?;
        conn.query_row(
            "SELECT id, repo_id, jira_account_id, jira_board_id, jira_issue_key, title, description, assignee, status, priority, source, require_plan, auto_start, auto_approve_plan, use_worktree, last_pipeline_error, last_pipeline_at, agent_profile_id, pr_url, pr_number, payload_json, created_at, updated_at FROM tasks WHERE id = ?1",
            [task_id],
            |row| {
                let payload_raw: String = row.get(20)?;
                Ok(TaskWithPayload {
                    id: row.get(0)?,
                    repo_id: row.get(1)?,
                    jira_account_id: row.get(2)?,
                    jira_board_id: row.get(3)?,
                    jira_issue_key: row.get(4)?,
                    title: row.get(5)?,
                    description: row.get(6)?,
                    assignee: row.get(7)?,
                    status: row.get(8)?,
                    priority: row.get(9)?,
                    source: row.get(10)?,
                    require_plan: row.get::<_, i64>(11)? != 0,
                    auto_start: row.get::<_, i64>(12)? != 0,
                    auto_approve_plan: row.get::<_, i64>(13)? != 0,
                    use_worktree: row.get::<_, i64>(14)? != 0,
                    last_pipeline_error: row.get(15)?,
                    last_pipeline_at: row.get(16)?,
                    agent_profile_id: row.get(17)?,
                    pr_url: row.get(18)?,
                    pr_number: row.get(19)?,
                    payload: serde_json::from_str(&payload_raw).unwrap_or(Value::Null),
                    created_at: row.get(21)?,
                    updated_at: row.get(22)?,
                })
            },
        )
        .optional()
        .map_err(anyhow::Error::from)
    }

    pub fn delete_task(&self, task_id: &str) -> Result<()> {
        let conn = self.connect()?;
        let deleted = conn.execute("DELETE FROM tasks WHERE id = ?1", [task_id])?;
        if deleted == 0 {
            anyhow::bail!("Task not found: {}", task_id);
        }
        Ok(())
    }

    pub fn create_manual_task(&self, payload: &CreateTaskPayload) -> Result<TaskWithPayload> {
        let conn = self.connect()?;
        let id = Uuid::new_v4().to_string();
        let ts = now_iso();
        let status = payload.status.as_deref().unwrap_or("To Do");
        let require_plan = payload.require_plan.unwrap_or(true);
        let auto_start = payload.auto_start.unwrap_or(false);
        let auto_approve_plan = payload.auto_approve_plan.unwrap_or(false);
        let use_worktree = payload.use_worktree.unwrap_or(true);

        // Generate LOCAL-N key
        let max_local: Option<String> = conn
            .query_row(
                "SELECT jira_issue_key FROM tasks WHERE jira_issue_key LIKE 'LOCAL-%' ORDER BY CAST(SUBSTR(jira_issue_key, 7) AS INTEGER) DESC LIMIT 1",
                [],
                |row| row.get(0),
            )
            .optional()?;
        let next_num = max_local
            .and_then(|k| k.strip_prefix("LOCAL-").and_then(|n| n.parse::<i64>().ok()))
            .unwrap_or(0)
            + 1;
        let issue_key = format!("LOCAL-{}", next_num);

        conn.execute(
            r#"INSERT INTO tasks (
                 id, repo_id, jira_account_id, jira_board_id, jira_issue_key, title,
                 description, assignee, status, priority, source, require_plan, auto_start,
                 auto_approve_plan, use_worktree, agent_profile_id, last_pipeline_error, last_pipeline_at, payload_json, created_at, updated_at
               ) VALUES (?1, ?2, NULL, NULL, ?3, ?4, ?5, NULL, ?6, ?7, 'manual', ?8, ?9, ?10, ?11, ?12, NULL, NULL, '{}', ?13, ?14)"#,
            params![
                id,
                payload.repo_id,
                issue_key,
                payload.title,
                payload.description,
                status,
                payload.priority,
                require_plan,
                auto_start,
                auto_approve_plan,
                use_worktree,
                payload.agent_profile_id,
                ts,
                ts,
            ],
        )?;
        self.get_task_by_id(&id)?
            .context("task missing after insert")
    }

    pub fn update_task_status(&self, task_id: &str, status: &str) -> Result<()> {
        let conn = self.connect()?;
        conn.execute(
            "UPDATE tasks SET status = ?1, updated_at = ?2 WHERE id = ?3",
            params![status, now_iso(), task_id],
        )?;
        Ok(())
    }

    pub fn update_task_pipeline_state(
        &self,
        task_id: &str,
        last_pipeline_error: Option<&str>,
    ) -> Result<()> {
        let conn = self.connect()?;
        conn.execute(
            "UPDATE tasks SET last_pipeline_error = ?1, last_pipeline_at = ?2, updated_at = ?2 WHERE id = ?3",
            params![last_pipeline_error, now_iso(), task_id],
        )?;
        Ok(())
    }

    pub fn update_task_details(
        &self,
        task_id: &str,
        title: &str,
        description: Option<&str>,
        priority: Option<&str>,
        require_plan: bool,
        auto_start: bool,
        auto_approve_plan: bool,
        use_worktree: bool,
        agent_profile_id: Option<&str>,
    ) -> Result<()> {
        let conn = self.connect()?;
        conn.execute(
            "UPDATE tasks SET title = ?1, description = ?2, priority = ?3, require_plan = ?4, auto_start = ?5, auto_approve_plan = ?6, use_worktree = ?7, agent_profile_id = ?8, updated_at = ?9 WHERE id = ?10",
            params![
                title,
                description,
                priority,
                require_plan,
                auto_start,
                auto_approve_plan,
                use_worktree,
                agent_profile_id,
                now_iso(),
                task_id
            ],
        )?;
        Ok(())
    }

    pub fn update_task_pr(&self, task_id: &str, pr_url: &str, pr_number: Option<i64>) -> Result<()> {
        let conn = self.connect()?;
        conn.execute(
            "UPDATE tasks SET pr_url = ?1, pr_number = ?2, updated_at = ?3 WHERE id = ?4",
            params![pr_url, pr_number, now_iso(), task_id],
        )?;
        Ok(())
    }
}

File: server-ts/src/db/tasks.ts
import { v4 as uuidv4 } from 'uuid';
import type {
  CreateTaskPayload,
  TaskWithPayload,
} from '../models.js';
import { Db, nowIso } from './index.js';

declare module './index.js' {
  interface Db {
    listTasksByRepo(repoId: string): TaskWithPayload[];
    getTaskById(taskId: string): TaskWithPayload | null;
    deleteTask(taskId: string): void;
    createManualTask(payload: CreateTaskPayload): TaskWithPayload;
    updateTaskStatus(taskId: string, status: string): void;
    updateTaskPipelineState(taskId: string, lastPipelineError?: string): void;
    updateTaskDetails(
      taskId: string,
      title: string,
      description: string | undefined,
      priority: string | undefined,
      requirePlan: boolean,
      autoStart: boolean,
      autoApprovePlan: boolean,
      useWorktree: boolean,
      carryDirtyState: boolean,
      agentProfileId?: string,
    ): void;
    updateTaskPr(taskId: string, prUrl: string, prNumber?: number): void;
    getNextQueueTask(repoId: string): TaskWithPayload | null;
    reorderTasks(taskIds: string[]): void;
  }
}

function rowToTask(row: any): TaskWithPayload {
  return {
    id: row.id,
    repo_id: row.repo_id,
    jira_account_id: row.jira_account_id,
    jira_board_id: row.jira_board_id,
    jira_issue_key: row.jira_issue_key,
    title: row.title,
    description: row.description,
    assignee: row.assignee,
    status: row.status,
    priority: row.priority,
    source: row.source,
    require_plan: !!row.require_plan,
    auto_start: !!row.auto_start,
    auto_approve_plan: !!row.auto_approve_plan,
    use_worktree: !!row.use_worktree,
    carry_dirty_state: !!row.carry_dirty_state,
    last_pipeline_error: row.last_pipeline_error,
    last_pipeline_at: row.last_pipeline_at,
    agent_profile_id: row.agent_profile_id,
    pr_url: row.pr_url,
    pr_number: row.pr_number,
    sort_order: row.sort_order ?? 0,
    payload: JSON.parse(row.payload_json || '{}'),
    created_at: row.created_at,
    updated_at: row.updated_at,
  };
}

const TASK_COLS =
  'id, repo_id, jira_account_id, jira_board_id, jira_issue_key, title, description, assignee, status, priority, source, require_plan, auto_start, auto_approve_plan, use_worktree, carry_dirty_state, last_pipeline_error, last_pipeline_at, agent_profile_id, pr_url, pr_number, sort_order, payload_json, created_at, updated_at';

Db.prototype.listTasksByRepo = function (repoId: string): TaskWithPayload[] {
  const db = this.connect();
    const rows = db
      .prepare(`SELECT ${TASK_COLS} FROM tasks WHERE repo_id = ? ORDER BY sort_order ASC, created_at ASC`)
      .all(repoId) as any[];
    return rows.map(rowToTask);
};

Db.prototype.getTaskById = function (taskId: string): TaskWithPayload | null {
  const db = this.connect();
    const row = db.prepare(`SELECT ${TASK_COLS} FROM tasks WHERE id = ?`).get(taskId) as
      | any
      | undefined;
    return row ? rowToTask(row) : null;
};

Db.prototype.deleteTask = function (taskId: string): void {
  const db = this.connect();
    const result = db.prepare('DELETE FROM tasks WHERE id = ?').run(taskId);
    if (Number(result.changes) === 0) {
      throw new Error(`Task not found: ${taskId}`);
    }
};

Db.prototype.createManualTask = function (payload: CreateTaskPayload): TaskWithPayload {
  const db = this.connect();
    const id = uuidv4();
    const ts = nowIso();
    const status = payload.status ?? 'To Do';
    const defaults = this.resolveTaskDefaults(payload.repoId);
    const requirePlan = payload.requirePlan ?? defaults.require_plan ?? true;
    const autoStart = payload.autoStart ?? defaults.auto_start ?? false;
    const autoApprovePlan = payload.autoApprovePlan ?? defaults.auto_approve_plan ?? false;
    const useWorktree = payload.useWorktree ?? defaults.use_worktree ?? true;
    const carryDirtyState = payload.carryDirtyState ?? defaults.carry_dirty_state ?? false;

    // Use current timestamp as sort_order so a newly created task always sorts
    // strictly after every existing task (existing values are small integers
    // from reorderTasks(); a ms-since-epoch value dwarfs them). A manual
    // reorder re-compresses the lane back to 0..N when the user drags.
    const sortOrder = Date.now();

    const maxLocal = db
      .prepare(
        "SELECT jira_issue_key FROM tasks WHERE jira_issue_key LIKE 'LOCAL-%' ORDER BY CAST(SUBSTR(jira_issue_key, 7) AS INTEGER) DESC LIMIT 1",
      )
      .get() as { jira_issue_key: string } | undefined;

    let nextNum = 1;
    if (maxLocal) {
      const prefix = 'LOCAL-';
      const numStr = maxLocal.jira_issue_key.slice(prefix.length);
      const parsed = parseInt(numStr, 10);
      if (!isNaN(parsed)) {
        nextNum = parsed + 1;
      }
    }
    const issueKey = `LOCAL-${nextNum}`;

    db.prepare(
      `INSERT INTO tasks (
         id, repo_id, jira_account_id, jira_board_id, jira_issue_key, title,
         description, assignee, status, priority, source, require_plan, auto_start,
         auto_approve_plan, use_worktree, carry_dirty_state, agent_profile_id, last_pipeline_error, last_pipeline_at, sort_order, payload_json, created_at, updated_at
       ) VALUES (?, ?, NULL, NULL, ?, ?, ?, NULL, ?, ?, 'manual', ?, ?, ?, ?, ?, ?, NULL, NULL, ?, '{}', ?, ?)`,
    ).run(
      id,
      payload.repoId,
      issueKey,
      payload.title,
      payload.description ?? null,
      status,
      payload.priority ?? null,
      requirePlan ? 1 : 0,
      autoStart ? 1 : 0,
      autoApprovePlan ? 1 : 0,
      useWorktree ? 1 : 0,
      carryDirtyState ? 1 : 0,
      payload.agentProfileId ?? null,
      sortOrder,
      ts,
      ts,
    );

    const row = db.prepare(`SELECT ${TASK_COLS} FROM tasks WHERE id = ?`).get(id) as any;
    return rowToTask(row);
};

Db.prototype.updateTaskStatus = function (taskId: string, status: string): void {
  const db = this.connect();
    db.prepare('UPDATE tasks SET status = ?, updated_at = ? WHERE id = ?').run(
      status,
      nowIso(),
      taskId,
    );
};

Db.prototype.updateTaskPipelineState = function (
  taskId: string,
  lastPipelineError?: string,
): void {
  const db = this.connect();
    const ts = nowIso();
    db.prepare(
      'UPDATE tasks SET last_pipeline_error = ?, last_pipeline_at = ?, updated_at = ? WHERE id = ?',
    ).run(lastPipelineError ?? null, ts, ts, taskId);
};

Db.prototype.updateTaskDetails = function (
  taskId: string,
  title: string,
  description: string | undefined,
  priority: string | undefined,
  requirePlan: boolean,
  autoStart: boolean,
  autoApprovePlan: boolean,
  useWorktree: boolean,
  carryDirtyState: boolean,
  agentProfileId?: string,
): void {
  const db = this.connect();
    db.prepare(
      'UPDATE tasks SET title = ?, description = ?, priority = ?, require_plan = ?, auto_start = ?, auto_approve_plan = ?, use_worktree = ?, carry_dirty_state = ?, agent_profile_id = ?, updated_at = ? WHERE id = ?',
    ).run(
      title,
      description ?? null,
      priority ?? null,
      requirePlan ? 1 : 0,
      autoStart ? 1 : 0,
      autoApprovePlan ? 1 : 0,
      useWorktree ? 1 : 0,
      carryDirtyState ? 1 : 0,
      agentProfileId ?? null,
      nowIso(),
      taskId,
    );
};

Db.prototype.updateTaskPr = function (
  taskId: string,
  prUrl: string,
  prNumber?: number,
): void {
  const db = this.connect();
    db.prepare('UPDATE tasks SET pr_url = ?, pr_number = ?, updated_at = ? WHERE id = ?').run(
      prUrl,
      prNumber ?? null,
      nowIso(),
      taskId,
    );
};

Db.prototype.getNextQueueTask = function (repoId: string): TaskWithPayload | null {
  const db = this.connect();
  const row = db
    .prepare(
      `SELECT ${TASK_COLS} FROM tasks
       WHERE repo_id = ? AND auto_start = 1 AND UPPER(TRIM(status)) IN ('TO DO', 'TODO')
       ORDER BY sort_order ASC LIMIT 1`,
    )
    .get(repoId) as any | undefined;
  return row ? rowToTask(row) : null;
};

Db.prototype.reorderTasks = function (taskIds: string[]): void {
  const ts = nowIso();
  const tx = this.transaction(() => {
    const db = this.connect();
    const stmt = db.prepare('UPDATE tasks SET sort_order = ?, updated_at = ? WHERE id = ?');
    for (let i = 0; i < taskIds.length; i++) {
      stmt.run(i, ts, taskIds[i]);
    }
  });
  tx();
};
